src/perf/bin/appmain.cpp

```diff
@@ -0,0 +1,204 @@
+/*++
+
+    Copyright (c) Microsoft Corporation.
+    Licensed under the MIT License.
+
+Abstract:
+
+    QUIC Perf Main execution runner.
+
+--*/
+
+#ifdef QUIC_CLOG
+#include "appmain.cpp.clog.h"
+#endif
+
+#define QUIC_TEST_APIS 1
+#include "quic_driver_main.h"
+#include "PerfHelpers.h"
+#include <quic_trace.h>
+
+#ifdef _WIN32
+
+//
+// Name of the driver service for quicperf.sys.
+// Must be defined before quic_driver_helpers.h is included
+//
+#define QUIC_DRIVER_NAME   "quicperf"
+
+#include <winioctl.h>
+#include "perfioctls.h"
+#include "quic_driver_helpers.h"
+
+#endif
+
+extern "C" _IRQL_requires_max_(PASSIVE_LEVEL) void QuicTraceRundown(void) { }
+
+int
+QuicUserMain(
+    _In_ int argc,
+    _In_reads_(argc) _Null_terminated_ char* argv[],
+    _In_ bool KeyboardWait
+    ) {
+    QUIC_EVENT StopEvent;
+    QuicEventInitialize(&StopEvent, true, false);
+
+    int RetVal = QuicMainStart(argc, argv, StopEvent);
+    if (RetVal != 0) {
+        return RetVal;
+    }
+    printf("Ready For Connections!\n\n");
+    fflush(stdout);
+    if (KeyboardWait) {
+        printf("Press enter to exit\n");
+        getchar();
+        QuicEventSet(StopEvent);
+    }
+    RetVal = QuicMainStop(0);
+    QuicEventUninitialize(StopEvent);
+    return RetVal;
+}
+
+#ifdef _WIN32
+
+int
+QuicKernelMain(
+    _In_ int argc,
+    _In_reads_(argc) _Null_terminated_ char* argv[],
+    _In_ bool KeyboardWait,
+    _In_ QUIC_SEC_CONFIG_PARAMS* SelfSignedParams
+    ) {
+    size_t TotalLength = 0;
+
+    // Get total length
+    for (int i = 0; i < argc; ++i) {
+        //
+        // Length of string, plus null terminator, plus length
+        //
+        TotalLength += strlen(argv[i]) + sizeof(size_t) + 1;
+    }
+
+    TotalLength += sizeof(TotalLength);
+
+    char* Data = static_cast<char*>(QuicAlloc(TotalLength));
```

Comment:
Do not call QuicAlloc. QUIC_ALLOC_NONPAGED instead.

Reply:
Or use new/free

Reply:
Ok. Just curious whats wrong with using QuicAlloc?

Reply:
It's not the cross-platform abstraction we use for allocations. The fact that Linux and Windows user modes happen to have a function called that is just coincidence and may change in the future.